### PR TITLE
Allow moving typing off master

### PR DESCRIPTION
**Pull request by erikjohnston**
A lot of this is various bits of rejigging, so probably worth looking at the commits.

### CODE REVIEW

File: synapse/replication/tcp/handler.py [hunk 1]
```diff
@@ -96,6 +97,14 @@ def __init__(self, hs):
 
                 continue
 
+            if isinstance(stream, (TypingStream)):
```
Comment (clokep) [inline synapse/replication/tcp/handler.py:100]:
```suggestion
            if isinstance(stream, TypingStream):
```

File: synapse/replication/tcp/streams/_base.py [hunk 2]
```diff
@@ -294,7 +294,8 @@ class TypingStream(Stream):
     def __init__(self, hs):
         typing_handler = hs.get_typing_handler()
 
-        if hs.config.worker_app is None:
+        writer_instance = hs.config.worker.writers.typing
+        if writer_instance == hs.get_instance_name():
             # on the master, query the typing handler
```
Comment (clokep) [inline synapse/replication/tcp/streams/_base.py:299]: I believe this comment is now out of date.

File: synapse/federation/federation_server.py [hunk 3]
```diff
@@ -809,66 +842,56 @@ def register_query_handler(
...
         if handler:
             return await handler(args)
 
-        return await self._get_query_client(query_type=query_type, args=args)
+        # Check if we can route it somewhere else that isn't us
+        if self._instance_name == "master":
+            return await self._get_query_client(query_type=query_type, args=args)
+
+        # Uh oh, no handler! Let's raise an exception so the request reutrns an
+        # error.
```
Comment (clokep) [inline synapse/federation/federation_server.py:895]:
```suggestion
        # Uh oh, no handler! Let's raise an exception so the request returns an
        # error.
```

File: synapse/handlers/typing.py [hunk 4]
```diff
@@ -39,48 +43,48 @@
...
-        hs.get_distributor().observe("user_left_room", self.user_left_room)
+        # map room IDs to serial numbers
+        self._room_serials = {}
+        # map room IDs to sets of users currently typing
+        self._room_typing = {}
 
-        self._member_typing_until = {}  # clock time we expect to stop
         self._member_last_federation_poke = {}
-
+        self.wheel_timer = WheelTimer(bucket_size=5000)
```
Comment (clokep) [inline synapse/handlers/typing.py:72]: We could just call `self._reset` here, but I suspect this code was all copy and pasted. 😄

Comment (erikjohnston) [reply synapse/handlers/typing.py:72]: Yeeeeeeeeeeeeeees, though I'm not a huge fan of declaring attributes in not `__init__` tbh :)

Comment (clokep) [reply synapse/handlers/typing.py:72]: Ah, that's true. I think type checkers don't like that. :)

**Pull request merged**